### PR TITLE
Use protobuf in networkdb core messages

### PR DESCRIPTION
**Pull request by mrjana**
Convert all networkdb core message types from go message types to
protobuf message types. This faciliates future modification of the
message structure without breaking backward compatibility.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

### CODE REVIEW

File: networkdb/networkdb.go [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package networkdb
 
+//go:generate protoc -I.:../Godeps/_workspace/src/github.com/gogo/protobuf  --gogo_out=import_path=github.com/docker/libnetwork/networkdb,Mgogoproto/gogo.proto=github.com/gogo/protobuf/gogoproto:. networkdb.proto
+
```
Comment (sanimej) [inline networkdb/networkdb.go]: @mrjana I am not that familiar with gogo protobuf. What does this `Mgogoproto/gogo.proto=github.com/gogo/protobuf/gogoproto` do ?

Comment (mrjana) [inline networkdb/networkdb.go]: It tells which go package to import in generated code for the corresponding import in .proto

**Comment by sanimej** [comment]
One general question on the protobuf usage. How do we handle deprecating/retiring a protofbuf field in golang ? I think other language bindings have APIs like has_<field> to check. Unless we have such checks the default value will be interested as a valid value for the deprecated fields.

**Comment by mrjana** [comment]
Newer version code will have to ignore deprecated values. The older version code should be able to understand a default value which is either a pretty good default or the default value should mean something is invalid. That is why you would see that the zero value in all the enum defines are actually invalid values.

**Comment by sanimej** [comment]
For fields that are not set go defaults might not be sufficient to indicate its invalid. For ex: if we decide to use a NodeID rather than NodeName in future, the default value of nil string for NodeName would be incorrect in handleNetworkEvent(). May be we can use the protobuf default option for such fields ?

**Comment by mrjana** [comment]
I don't know even if HasField is available it's going to help in that case because the old version cannot get access to `NodeID` which means if it is mandatory to have NodeName we cannot remove it. As long as we check if NodeName of empty string and mark it invalid we should be fine.

**Comment by sanimej** [comment]
yes, NodeName can't be removed. But for this and other fields we have to add a check to make sure the default values doesn't get processed. It can be done in a separate PR though.

**Comment by sanimej** [comment]
LGTM

**Pull request merged**